react-renderer/src/Optimizer.ts

```diff
@@ -114,53 +107,69 @@ export const evalEnergyOn = (state: State) => {
     );
 
     // Printing to check for NaNs
-    console.log(
-      zip(
-        constrEngs.map((e) => e.toString()),
-        constrEvaled.map((c) => c.name)
-      )
-    );
+    // console.log(
+    //   zip(
+    //     constrEngs.map((e) => e.toString()),
+    //     constrEvaled.map((c) => c.name)
+    //   ),
+    //   varyingMap
+    // );
 
     const objEng: Tensor =
       objEngs.length === 0 ? differentiable(0) : stack(objEngs).sum();
     const constrEng: Tensor =
       constrEngs.length === 0 ? differentiable(0) : stack(constrEngs).sum();
+    const overallEng = objEng.add(
+      constrEng.mul(scalar(constraintWeight * state.paramsr.weight))
+    );
 
-    const dummyVal = stack(varyingValuesTF)
-      .square()
-      .sum();
-    return dummyVal.add(constrEng);
-
-    // return constrEng.add(
-    //   stack(varyingValuesTF)
-    //     .sum()
-    //     .mul(0)
-    // );
-
-    // return objEng.add(
-    //   constrEng.mul(tf.scalar(constraintWeight * state.paramsr.weight))
-    // );
+    // NOTE: the current version of tfjs requires all input variables to have gradients (i.e. actually involved when computing the overall energy). See https://github.com/tensorflow/tfjs-core/blob/8c2d9e05643988fa7f4575c30a5ad3e732d189b2/tfjs-core/src/engine.ts#L726
+    // HACK: therefore, we try to work around it by using all varying values without affecting the value and gradients of the energy function
+    const dummyVal = stack(varyingValuesTF).sum();
+    return overallEng.sub(dummyVal).add(dummyVal);
```

Comment:
I know this is an interim hack until we can fork tfjs, but these lines might make gradient computation slower by adding extra operations that involve all variables. Not sure how much slower since I don't know how the AD is implemented.

Reply:
Btw, is there any floating point danger here, where the add/sub will actually create a tiny gradient? Could that be avoided by adding zero instead, multiplying by 1, etc?

Reply:
Yeah, my suggestion is to do something like `overallEng + sum(vector_of_zeros `dot` varyingValuesTF)` (maybe with a `stack` in there). Multiplying by zero is a very safe thing to do with floats!

Reply:
Will do!